File: server/proliferate/server/cloud/workflows/contracts/schema_profile.py
"""The v1 emit JSON Schema profile validator (feature spec §6.2).

V1 accepts only the vocabulary implemented in all three contract languages;
unsupported keywords are rejected at save/compile rather than ignored. This
mirror is byte-for-byte equivalent in behavior to the Rust and TypeScript
validators over the shared golden valid/invalid fixtures.
"""

from __future__ import annotations

import re
from typing import Any

DRAFT_2020_12 = "https://json-schema.org/draft/2020-12/schema"

ALLOWED_KEYWORDS = frozenset(
    {
        "$schema",
        "type",
        "properties",
        "required",
        "additionalProperties",
        "items",
        "enum",
        "const",
        "minimum",
        "maximum",
        "minLength",
        "maxLength",
        "minItems",
        "maxItems",
        "title",
        "description",
        "default",
    }
)

_JSON_TYPES = frozenset({"object", "array", "string", "number", "integer", "boolean", "null"})

_ASCII_IDENTIFIER = re.compile(r"^[A-Za-z_][A-Za-z0-9_]*$")


class SchemaProfileError(ValueError):
    """A schema violates the v1 emit profile. ``code`` is the stable reason."""

    def __init__(self, code: str, message: str) -> None:
        super().__init__(f"{code}: {message}")
        self.code = code


def _validate_type(node_type: Any) -> None:  # noqa: ANN401 - validates untrusted JSON
    if isinstance(node_type, str):
        if node_type not in _JSON_TYPES:
            raise SchemaProfileError("invalid_type", f"unknown type {node_type!r}")
        return
    if isinstance(node_type, list):
        if len(node_type) != 2 or node_type[1] != "null":
            raise SchemaProfileError(
                "invalid_type",
                'a type union must be a two-item [TYPE, "null"]',
            )
        head = node_type[0]
        if not isinstance(head, str) or head not in _JSON_TYPES or head == "null":
            raise SchemaProfileError("invalid_type", "invalid union head type")
        return
    raise SchemaProfileError("invalid_type", "type must be a string or a two-item union")


def _validate_node(node: Any, *, is_root: bool) -> None:  # noqa: ANN401 - validates untrusted JSON
    if not isinstance(node, dict):
        raise SchemaProfileError("node_not_object", "schema node must be an object")

    for key in node:
        if key not in ALLOWED_KEYWORDS:
            raise SchemaProfileError("unsupported_keyword", f"keyword {key!r} is not permitted")

    schema_dialect = node.get("$schema")
    if is_root:
        if schema_dialect is not None and schema_dialect != DRAFT_2020_12:
            raise SchemaProfileError("wrong_dialect", "root $schema must be draft 2020-12")
        if node.get("type") != "object":
            raise SchemaProfileError("root_type_not_object", "emit root must be type object")
    elif schema_dialect is not None:
        raise SchemaProfileError("unsupported_keyword", "$schema only allowed at the root")

    if "type" in node:
        _validate_type(node["type"])

    for numeric_key in ("minimum", "maximum"):
        if numeric_key in node and not isinstance(node[numeric_key], (int, float)):
            raise SchemaProfileError("invalid_bound", f"{numeric_key} must be a finite number")
    for int_key in ("minLength", "maxLength", "minItems", "maxItems"):
        if int_key in node and (
            not isinstance(node[int_key], int) or isinstance(node[int_key], bool)
        ):
            raise SchemaProfileError("invalid_bound", f"{int_key} must be an integer")

    if "enum" in node and not isinstance(node["enum"], list):
        raise SchemaProfileError("invalid_enum", "enum must be an array")

    properties = node.get("properties")
    if properties is not None:
        if not isinstance(properties, dict):
            raise SchemaProfileError("invalid_properties", "properties must be an object")
        for prop_name, prop_schema in properties.items():
            if is_root and not _ASCII_IDENTIFIER.match(prop_name):
                raise SchemaProfileError(
                    "property_name_not_ascii_identifier",
                    f"top-level property {prop_name!r} is not an ASCII identifier",
                )
            _validate_node(prop_schema, is_root=False)

    required = node.get("required")
    if required is not None and not (
        isinstance(required, list) and all(isinstance(r, str) for r in required)
    ):
        raise SchemaProfileError("invalid_required", "required must be an array of strings")

    items = node.get("items")
    if items is not None:
        _validate_node(items, is_root=False)


def validate_schema_profile(document: Any) -> None:  # noqa: ANN401 - public JSON boundary
    """Raise ``SchemaProfileError`` if ``document`` is not a valid v1 emit schema."""

    _validate_node(document, is_root=True)

File: apps/packages/product-domain/src/workflows/contracts/schema-profile.ts
/**
 * The v1 emit JSON Schema profile validator (feature spec §6.2).
 *
 * V1 accepts only the vocabulary implemented in all three contract languages;
 * unsupported keywords are rejected at save/compile. Behavior is byte-for-byte
 * equivalent to the Python `schema_profile.py` validator over the shared golden
 * valid/invalid fixtures.
 */

export const DRAFT_2020_12 = "https://json-schema.org/draft/2020-12/schema";

const ALLOWED_KEYWORDS = new Set([
  "$schema",
  "type",
  "properties",
  "required",
  "additionalProperties",
  "items",
  "enum",
  "const",
  "minimum",
  "maximum",
  "minLength",
  "maxLength",
  "minItems",
  "maxItems",
  "title",
  "description",
  "default",
]);

const JSON_TYPES = new Set(["object", "array", "string", "number", "integer", "boolean", "null"]);

const ASCII_IDENTIFIER = /^[A-Za-z_][A-Za-z0-9_]*$/;

export class SchemaProfileError extends Error {
  readonly code: string;

  constructor(code: string, message: string) {
    super(`${code}: ${message}`);
    this.code = code;
    this.name = "SchemaProfileError";
  }
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function validateType(nodeType: unknown): void {
  if (typeof nodeType === "string") {
    if (!JSON_TYPES.has(nodeType)) {
      throw new SchemaProfileError("invalid_type", `unknown type ${nodeType}`);
    }
    return;
  }
  if (Array.isArray(nodeType)) {
    if (nodeType.length !== 2 || nodeType[1] !== "null") {
      throw new SchemaProfileError("invalid_type", 'a type union must be a two-item [TYPE, "null"]');
    }
    const head = nodeType[0];
    if (typeof head !== "string" || !JSON_TYPES.has(head) || head === "null") {
      throw new SchemaProfileError("invalid_type", "invalid union head type");
    }
    return;
  }
  throw new SchemaProfileError("invalid_type", "type must be a string or a two-item union");
}

function validateNode(node: unknown, isRoot: boolean): void {
  if (!isPlainObject(node)) {
    throw new SchemaProfileError("node_not_object", "schema node must be an object");
  }

  for (const key of Object.keys(node)) {
    if (!ALLOWED_KEYWORDS.has(key)) {
      throw new SchemaProfileError("unsupported_keyword", `keyword ${key} is not permitted`);
    }
  }

  const dialect = node["$schema"];
  if (isRoot) {
    if (dialect !== undefined && dialect !== DRAFT_2020_12) {
      throw new SchemaProfileError("wrong_dialect", "root $schema must be draft 2020-12");
    }
    if (node["type"] !== "object") {
      throw new SchemaProfileError("root_type_not_object", "emit root must be type object");
    }
  } else if (dialect !== undefined) {
    throw new SchemaProfileError("unsupported_keyword", "$schema only allowed at the root");
  }

  if ("type" in node) {
    validateType(node["type"]);
  }

  for (const numericKey of ["minimum", "maximum"]) {
    if (numericKey in node && typeof node[numericKey] !== "number") {
      throw new SchemaProfileError("invalid_bound", `${numericKey} must be a finite number`);
    }
  }
  for (const intKey of ["minLength", "maxLength", "minItems", "maxItems"]) {
    if (intKey in node && !Number.isInteger(node[intKey])) {
      throw new SchemaProfileError("invalid_bound", `${intKey} must be an integer`);
    }
  }

  if ("enum" in node && !Array.isArray(node["enum"])) {
    throw new SchemaProfileError("invalid_enum", "enum must be an array");
  }

  const properties = node["properties"];
  if (properties !== undefined) {
    if (!isPlainObject(properties)) {
      throw new SchemaProfileError("invalid_properties", "properties must be an object");
    }
    for (const [propName, propSchema] of Object.entries(properties)) {
      if (isRoot && !ASCII_IDENTIFIER.test(propName)) {
        throw new SchemaProfileError(
          "property_name_not_ascii_identifier",
          `top-level property ${propName} is not an ASCII identifier`,
        );
      }
      validateNode(propSchema, false);
    }
  }

  const required = node["required"];
  if (
    required !== undefined &&
    !(Array.isArray(required) && required.every((r) => typeof r === "string"))
  ) {
    throw new SchemaProfileError("invalid_required", "required must be an array of strings");
  }

  const items = node["items"];
  if (items !== undefined) {
    validateNode(items, false);
  }
}

/** Throws SchemaProfileError if `document` is not a valid v1 emit schema. */
export function validateSchemaProfile(document: unknown): void {
  validateNode(document, true);
}
